README.md
# datastore-backend

A backend to store and retrieve data for [*entities*](#entity).

## Setup

### Prerequisites

Make sure you have the following installed:

* MongoDB

## API

Every entity can save a [*data sets*](#data-set). The data sets can only be read by [*authorized parties*](#authorized-party). New data sets can only be created by authorized parties. As well can changes to existing data sets only be made by authorized parties.

The data store must be accessed by HTTPS.

### Versioning

The current version of the API is ``v1``. **All URLs must be prefixed
with ``/v1``!**

### Create a new data set

#### Request

**POST** ``/:UUID:``

##### Parameters

- **UUID** [URL]: The UUID of the entity the data set should be created for

##### Body
JSON encoded hash to be stored.

#### Response

##### Body

JSON encoded hash like this:
```{"uuid": "the-uuid-of-the-document", "data": "{\"some\":
\"data\"}"}```

### Change a data set

#### Request

**PUT** ``/:UUID:/:key:``

##### Parameters

- **UUID** [URL] [REQUIRED]: The UUID of the entity the data set should be created for
- **key** [URL] [OPTIONAL]: Makes it possible to change only a subset of
  the stored data. The key determines which key in the hash is changed.
You can nest the key by separating it with a ``/`` (slash). If no key is
specified the whole data set will be replaced.

##### Body
JSON encoded hash to be stored.

#### Response

##### Body

JSON encoded hash like this:
```{"uuid": "the-uuid-of-the-document", "data": "{\"some\":
\"data\"}"}```

The response always contains the whole data set no matter if only a sub
set of the data set was changed (via a ``key`` parameter).

#### Example

##### Changing the whole data set

Send a ``PUT`` to ``/some-uuid`` with a JSON body like this
``{"some": "value"}``.

##### Changing a subset of a data set

If you have a data set with the uuid ``some-uuid`` that looks like this:

```javascript
{
  "colors": {
    "red": 100,
    "yellow": 12
  },
  "name": "A cool thing"
}
```

you could change it's name like this:

Send a ``PUT`` to ``/some-uuid/name`` with a JSON body like this
``{"name": "Another name"}``.

The data set would then look like this:

```javascript
{
  "colors": {
    "red": 100,
    "yellow": 12
  },
  "name": "Another name"
}
```

You could change the count of the color red like this:

Send a ``PUT`` to ``/some-uuid/colors/red`` with a JSON body like
this ``{"red": 99}``.

The data set would then look like this:

```javascript
{
  "colors": {
    "red": 99,
    "yellow": 12
  },
  "name": "A cool thing"
}
```

You could also change the count of all colors like this:

Send a ``PUT`` to ``/some-uuid/colors`` with a JSON body like
this ``{"colors": {"red": 92, "yellow": 10, "purple": 30}}``.

The data set would then look like this:

```javascript
{
  "colors": {
    "red": 92,
    "yellow": 10,
    "purple": 30
  },
  "name": "A cool thing"
}
```

It is also possible to add arbitrary nested hashed to the data set that
does not exist yet like this:

Send a ``PUT`` to ``/some-uuid/some/strange/thing`` with a JSON
body like this ``{"thing": {"yay": "YO"}}``.

The data set would then look like this:


```javascript
{
  "colors": {
    "red": 100,
    "yellow": 12
  },
  "name": "A cool thing",
  "some": {
    "strange": {
      "thing": {
        "yay": "YO"
      }
    }
  }
}
```

### Retrieve a data set

#### Request

**GET** ``/:UUID:``

##### Parameters

- **UUID** [URL] [REQUIRED]: The UUID of the entity the data set should be created for

##### Body
Empty

#### Response

##### Body

JSON encoded hash like this:
```{"uuid": "the-uuid-of-the-document", "data": "{\"some\":
\"data\"}"}```

### Error case response

HTTP status codes used:

* **401** Request needs to be authenticated. Either there were no credentials present in the request or the given credentials are invalid.
* **403** The authenticated party is not authorized to carry out the action requested or a validation error occured
* **404** No data set found for the given UUID

#### Body

Reason for the error encoded on JSON. Example:

```javascript
{
  error: 'Priviliges insufficient.'
}
```

### Retrieve a batch of data sets

Batch requests may only be issued by authorized clients of the type ``app``.

#### Request

**GET** ``/batch``

##### Body

A JSON encoded object.

##### Parameters

**uuids**: An array of UUIDs to retrieve.

#### Response

##### Body

JSON encoded hash of the requested data sets like this:

```javascript
{
  "some-uuid": {
    "uuid": "some-uuid",
    "data": {"some": "data"}
  },
  "some-other-uuid": {
    "uuid": "some-other-uuid",
    "data": {"some-more": "different data"}
  }
}
```

If a data set can not be found it will not be present in the response hash.

## Glosary

### Entity

An entity is anything that is described by a 128bit UUID as described in [RFC 4122](http://www.ietf.org/rfc/rfc4122.txt). Examples for entities are players but also games or developers.

### Authorized Party

Each request has to be authenticated by an OAuth token complying with the Quarter Spiral authentication guidelines. An authorized party for any entity is either an authorized client with the type ``user`` and the same UUID as the entity or any authorized client with the type ``app``.

### Data Set

A data set is a hash. Keys must be strings. Data sets support the following data types for their values:

* Null
* Boolean values (true/false)
* Integers
* Floating point numbers
* Strings
* Arrays (of any allowed data types)
* Data sets

Dates and times should be represented by numeric timestamps.


CHANGELOG.md
# 0.0.23 / 2013-08-12

* Fixed sentry app id

# 0.0.22 / 2013-08-12

* Fixes service client request tracking

# 0.0.21 / 2013-08-12

* Adds getsentry.com exception tracking
* Adds request id tracker
* Bumps client dependencies

# 0.0.20 / 2013-04-17

* Adds caching auth-client

# 0.0.19 / 2013-04-02

* Adds authorization to the datastore

# 0.0.18

* Fixes small bug in batch UUID retrieval

# 0.0.17

* Fixes Ruby version on Heroku

# 0.0.16

* Updates depencies

# 0.0.15

* Updates json

# 0.0.14

* Adds endpoint for batch retrieval of data sets

# 0.0.13

* Improve Newrelic instrumenting

# 0.0.12

* Adds Newrelic monitoring and ping middleware

# 0.0.11

* Moves to a fixed grape gem hosted on our own gem server instead of a git link

# 0.0.10

* Refactors API and dismisses the public/private scopes completely

# 0.0.9

* Relaxes dependency on auth-client

# 0.0.8

* ``OPTIONS`` requests don't need to be authenticated anymore

# 0.0.7

* Secures all requests with OAuth2

# 0.0.6

* Adds thin to the main bundle to run on heroku

# 0.0.5

* Makes it possible to change parts of data sets
* Hardens the API for the use behind Rack::Lint
* A lot of ``metaserver`` related improvements

# 0.0.4

* Fixes bug that prevented the server from booting

# 0.0.3

* Makes it possible to create data sets without specifying a UUID. A new
  one will be created for you.
* Adds the UUID of a data set to each response.

# 0.0.2

The start.


spec/spec_helper.rb
ENV['RACK_ENV'] ||= 'test'

Bundler.require

require 'rspec'
require 'datastore-backend'
require 'auth-backend'

require 'rack/client'
require 'json'
require 'uuid'

Dir[File.expand_path('../support/*.rb', __FILE__)].each {|f| require f}

RSpec.configure do |config|
end

class AuthenticationInjector
  def self.token=(token)
    @token = token
  end

  def self.token
    @token
  end

  def self.reset!
    @token = nil
  end

  def initialize(app)
    @app = app
  end

  def call(env)
    if token = self.class.token
      env['HTTP_AUTHORIZATION'] = "Bearer #{token}"
    end

    @app.call(env)
  end
end

ENV['QS_AUTH_BACKEND_URL'] = 'http://auth-backend.dev'

API_APP  = Datastore::Backend::API.new
AUTH_APP = Auth::Backend::App.new(test: true)

module Auth
  class Client
    alias raw_initialize initialize
    def initialize(url, options = {})
      raw_initialize(url, options.merge(adapter: [:rack, AUTH_APP]))
    end
  end
end

require 'auth-backend/test_helpers'
AUTH_HELPERS = Auth::Backend::TestHelpers.new(AUTH_APP)


spec/request/v1_spec.rb
require_relative '../spec_helper'

include Datastore::Backend

client = Rack::Client.new {
  use AuthenticationInjector

  run API_APP
}

sample_data = {
  'null' => nil,
  'boolean_true' => true,
  'boolean_false' => false,
  'integer' => 123,
  'float' => 123.456,
  'string' => 'some cool string',
  'array' => [123, 'Hallo', [1,2,3], false],
  'set' => {'key' => 'value', 'second' => 'value2', 'third' => [7, 'yip', {'inner'  => 'hash'}]}
}

token = AUTH_HELPERS.get_token

app = AUTH_HELPERS.create_app!
app_token = AUTH_HELPERS.get_app_token(app[:id], app[:secret])

describe "Datastore::Backend API" do
  before do
    @entity1 = UUID.new.generate
    @entity2 = UUID.new.generate
    AuthenticationInjector.reset!
  end

  after do
    DataSet.where(entity: @entity1).destroy
    DataSet.where(entity: @entity2).destroy
  end

  describe "data sets" do
    it "can't access the api unauthenticated" do
      client.post("/v1/#{@entity1}", {}, JSON.dump(sample_data)).status.should eq 403

      response = client.get("/v1/#{@entity1}")
      response.status.should eq 403
      JSON.parse(response.body).should eq('error' => 'Unauthenticated')

      client.put("/v1/#{@entity1}", {}, JSON.dump({test: 'yes'})).status.should eq 403
    end

    describe "authenticated" do
      before do
        AuthenticationInjector.token = app_token
      end

      after do
        AuthenticationInjector.reset!
      end

      it "not existing set returns 404" do
        response = client.get("/v1/#{@entity1}")

        response.status.should eq 404
      end

      it "can write data sets" do
        response = client.post("/v1/#{@entity1}", {}, JSON.dump(sample_data))

        response.status.should eq 201
        response_matches(response, @entity1, sample_data).should be true
      end

      it "cannot create data sets without an UUID" do
        response = client.post("/v1/", {}, JSON.dump(sample_data))
        response.status.should eq 405
        response.body.empty?.should be_true
      end

      it "can read data set after writing it" do
        client.post("/v1/#{@entity1}", {}, JSON.dump(sample_data))
        response = client.get("/v1/#{@entity1}")

        response.status.should eq 200
        response_matches(response, @entity1, sample_data).should be true
      end

      it "can retrieve batches of data sets" do
        client.post("/v1/#{@entity1}", {}, JSON.dump(sample_data))

        sample_data_2 = {"bla" => "blub", "yeah" => false}
        client.post("/v1/#{@entity2}", {}, JSON.dump(sample_data_2))

        not_existing_entity = @entity1.reverse
        response = client.get("/v1/batch?uuids=#{CGI.escape(JSON.dump([@entity1, @entity2, not_existing_entity]))}")

        response.status.should eq 200
        JSON.parse(response.body).should eq(
          @entity1 => {"uuid" => @entity1, "data" => sample_data},
          @entity2 => {"uuid" => @entity2, "data" => sample_data_2}
        )
      end

      it "retrieves an empty object when retrieving a batch with an empty uuids array" do
        response = client.get("/v1/batch?")
        response.status.should eq 200
        JSON.parse(response.body).should eq({})
      end

      it "errors out when creating a second data set for the same entity" do
        client.post("/v1/#{@entity1}", {}, JSON.dump(sample_data))
        response = client.post("/v1/#{@entity1}", {}, JSON.dump({}))

        response.status.should eq 403
        JSON.parse(response.body)['error'].should_not be_empty

        response = client.post("/v1/#{@entity2}", {}, JSON.dump({test: 'yes'}))
        response.status.should eq 201

        response = client.get("/v1/#{@entity1}")
        response_matches(response, @entity1, sample_data).should be true

        response = client.get("/v1/#{@entity2}")
        response_matches(response, @entity2, {'test' => 'yes'}).should be true
      end

      it "can change a data set" do
        client.post("/v1/#{@entity1}", {}, JSON.dump(sample_data))
        response = client.put("/v1/#{@entity1}", {}, JSON.dump({test: 'yes'}))

        response.status.should eq 200
        response_matches(response, @entity1, {'test' => 'yes'}).should be true
        response = client.get("/v1/#{@entity1}")
        response_matches(response, @entity1, {'test' => 'yes'}).should be true
      end

      describe "changing only some keys of a data set" do
        before do
          client.post("/v1/#{@entity1}", {}, JSON.dump(sample_data))
        end

        after do
          response_matches(@response, @entity1, @expected_data).should be true
          response = client.get("/v1/#{@entity1}")
          response_matches(response, @entity1, @expected_data).should be true
        end

        it "works with simple values" do
          @response = client.put("/v1/#{@entity1}/integer", {}, JSON.dump({'integer' => 456}))
          @expected_data = sample_data.merge('integer' => 456)
        end

        it "works with whole hashes" do
          @response = client.put("/v1/#{@entity1}/set", {}, JSON.dump({'set' => {'bla' => 'blub'}}))
          @expected_data = sample_data.merge('set' => {'bla' => 'blub'})
        end

        it "works with sub hashes" do
          @response = client.put("/v1/#{@entity1}/set/key", {}, JSON.dump({'key' => 'blub'}))
          @expected_data = sample_data.merge('set' => sample_data['set'].merge('key' => 'blub'))
        end

        it "creats non existing keys in existing hashes" do
          @response = client.put("/v1/#{@entity1}/set/newkey", {}, JSON.dump({'newkey' => 'blub'}))
          @expected_data = sample_data.merge('set' => sample_data['set'].merge('newkey' => 'blub'))
        end

        it "creates whole hashes" do
          @response = client.put("/v1/#{@entity1}/new/some/key", {}, JSON.dump({'key' => 'blub'}))
          @expected_data = sample_data.merge('new' => {'some' => {'key' => 'blub'}})
        end
      end

      it "errors out when changing a non existing data set" do
        response = client.put("/v1/#{@entity1}", {}, JSON.dump({test: 'yes'}))
        response.status.should eq 404
        JSON.parse(response.body).should eq({'error' => 'Not found'})
      end
    end
  end
end


spec/request/authorization_spec.rb
require_relative '../spec_helper'
require 'json'
require 'uuid'

include Datastore::Backend

require_relative '../spec_helper'

include Datastore::Backend

client = Rack::Client.new {
  use AuthenticationInjector

  run API_APP
}

user = AUTH_HELPERS.user_data['uuid']
token = AUTH_HELPERS.get_token

app = AUTH_HELPERS.create_app!
app_token = AUTH_HELPERS.get_app_token(app[:id], app[:secret])

describe "Authorization of the API" do
  describe "for users" do
    before do
      AuthenticationInjector.token = token
    end

    after do
      AuthenticationInjector.reset!
    end

    describe "with own data" do
      before do
        @set = {"test" => 'set'}
        DataSet.create!(entity: user, payload: @set)
      end

      after do
        DataSet.where(entity: user).destroy
      end

      it "allows reads" do
        response = client.get("/v1/#{user}")
        JSON.parse(response.body)['data'].should eq @set
      end

      it "allows creation" do
        DataSet.where(entity: user).destroy
        new_set = {"new" => "setto"}
        response = client.post("/v1/#{user}", {}, JSON.dump(new_set))
        JSON.parse(response.body)['data'].should eq new_set

        response = client.get("/v1/#{user}")
        JSON.parse(response.body)['data'].should eq new_set
      end

      it "allows updates" do
        new_set = {"new" => "setto"}
        response = client.put("/v1/#{user}", {}, JSON.dump(new_set))
        JSON.parse(response.body)['data'].should eq new_set

        response = client.get("/v1/#{user}")
        JSON.parse(response.body)['data'].should eq new_set
      end

      it "prevents batch requests" do
        response = client.get("/v1/batch?uuids[]=#{user}", {})
        response.status.should eq 403
        JSON.parse(response.body)['data'].should_not eq @set
      end
    end

    describe "with other data" do
      before do
        @entity = UUID.new.generate
        @set = {"test" => 'set'}
        DataSet.create!(entity: @entity, payload: @set)
      end

      after do
        DataSet.where(entity: @entity).destroy
      end

      it "prevents reads" do
        response = client.get("/v1/#{@entity}")
        response.status.should eq 403
        JSON.parse(response.body)['data'].should_not eq @set
      end

      it "prevents creation" do
        DataSet.where(entity: @entity).destroy
        new_set = {"new" => "setto"}
        response = client.post("/v1/#{@entity}", {}, JSON.dump(new_set))
        response.status.should eq 403
        JSON.parse(response.body)['data'].should_not eq new_set

        response = client.get("/v1/#{@entity}")
        response.status.should eq 403
        JSON.parse(response.body)['data'].should_not eq new_set
      end

      it "prevents updates" do
        new_set = {"new" => "setto"}
        response = client.put("/v1/#{@entity}", {}, JSON.dump(new_set))
        response.status.should eq 403
        JSON.parse(response.body)['data'].should_not eq new_set

        response = client.get("/v1/#{@entity}")
        response.status.should eq 403
        JSON.parse(response.body)['data'].should_not eq new_set
      end
    end
  end

  describe "for a client with system privileges" do
    describe "with any data" do
      before do
        AuthenticationInjector.token = app_token
        @entity = UUID.new.generate
        @set = {"test" => 'set'}
        DataSet.create!(entity: @entity, payload: @set)
      end

      after do
        DataSet.where(entity: @entity).destroy
      end

      it "allows reads" do
        response = client.get("/v1/#{@entity}")
        JSON.parse(response.body)['data'].should eq @set
      end

      it "allows creation" do
        DataSet.where(entity: @entity).destroy
        new_set = {"new" => "setto"}
        response = client.post("/v1/#{@entity}", {}, JSON.dump(new_set))
        JSON.parse(response.body)['data'].should eq new_set

        response = client.get("/v1/#{@entity}")
        JSON.parse(response.body)['data'].should eq new_set
      end

      it "allows updates" do
        new_set = {"new" => "setto"}
        response = client.put("/v1/#{@entity}", {}, JSON.dump(new_set))
        JSON.parse(response.body)['data'].should eq new_set

        response = client.get("/v1/#{@entity}")
        JSON.parse(response.body)['data'].should eq new_set
      end

      it "allows batch requests" do
        entity2 = UUID.new.generate
        set2 = {"new" => "setto"}
        DataSet.create!(entity: entity2, payload: set2)

        response = client.get("/v1/batch?uuids[]=#{@entity}&uuids[]=#{entity2}", {})
        response.status.should eq 200
        JSON.parse(response.body).should eq(@entity => {"uuid" => @entity, "data" => @set}, entity2 => {"uuid" => entity2, "data" => set2})

        DataSet.where(entity: entity2).destroy
      end
    end
  end
end


lib/datastore-backend/data_set.rb
require 'mongoid'

module Datastore::Backend
  class DataSet
    include Mongoid::Document

    field :entity,  type: String
    field :payload, type: Hash

    validates :entity, uniqueness: true

    index({entity: 1}, unique: true)

    def partial_update_attributes(key, payload)
      new_data = {}

      if self[key]
        data = payload[key]
        new_data = {key => data}
      else
        keys = key.split('/')
        data = payload[keys.last]

        current_sets_raw_payload = Hash[self.payload.to_a]
        new_data = deep_hash(keys, current_sets_raw_payload, data)
      end

      update_attributes(payload: new_data)
    end

    protected
    def deep_hash(keys, hash, data)
      key = keys.shift
      hash[key] = {} unless hash[key]
      if keys.empty?
        hash[key] = data
      else
        deep_hash(keys, hash[key], data)
      end

      hash
    end
  end
end


lib/datastore-backend/api.rb
require 'auth-client'
require 'grape'
require 'grape_newrelic'
require 'uuid'

module Datastore::Backend
  class API < ::Grape::API
    use GrapeNewrelic::Instrumenter
    version 'v1', :using => :path, :vendor => 'qs'

    format :json
    default_format :json

    rescue_from Mongoid::Errors::Validations
    rescue_from Mongoid::Errors::DocumentNotFound do
      Rack::Response.new([ JSON.dump({error: 'Not found'}) ], 404, { "Content-type" => "application/json" }).finish
    end
    default_error_formatter :json

    helpers do
      def payload
        return @payload if @payload

        body = request.body
        body =body.read if body.respond_to?(:read)

        @payload = JSON.parse(body)
      end

      def response_from_set(set)
        {uuid: set.entity, data: set.payload}
      end

      def create_set(uuid)
        set = DataSet.create!(entity: uuid, payload: payload)
        response_from_set set
      end

      def connection
        @connection ||= Connection.create
      end

      def own_data?(uuid)
        @token_owner['uuid'] == uuid
      end

      def system_level_privileges?
        @token_owner['type'] == 'app'
      end

      def is_authorized_to_access?(uuid)
        system_level_privileges? || own_data?(uuid)
      end

      def prevent_access!
        error!('Unauthenticated', 403)
      end

      def uuid
        params[:uuid]
      end
    end

    before do
      header('Access-Control-Allow-Origin', request.env['HTTP_ORIGIN'] || '*')

      unless request.env['REQUEST_METHOD'] == 'OPTIONS'
        error!('Unauthenticated', 403) unless request.env['HTTP_AUTHORIZATION']
        token = request.env['HTTP_AUTHORIZATION'].gsub(/^Bearer\s+/, '')
        @token_owner = connection.auth.token_owner(token)
        prevent_access! unless @token_owner
      end
    end

    get "version" do
      api.version
    end

    options '/' do
      header('Access-Control-Allow-Headers', '')
      header('Access-Control-Allow-Methods', '')
      ""
    end

    options '/:uuid' do
      header('Access-Control-Allow-Headers', 'origin, x-requested-with, content-type, accept, authorization')
      header('Access-Control-Allow-Methods', 'GET,PUT,OPTIONS, POST')
      header('Access-Control-Max-Age', '1728000')
      ""
    end

    get '/batch' do
      prevent_access! unless system_level_privileges?

      uuids = params[:uuids] || []
      uuids = JSON.parse(uuids) if uuids.kind_of?(String)

      sets = DataSet.where(entity: {'$in' => uuids})

      Hash[sets.map {|set| [set.entity, response_from_set(set)]}]
    end

    get '/:uuid' do
      prevent_access! unless is_authorized_to_access?(uuid)

      set = DataSet.where(entity: uuid).first
      raise Mongoid::Errors::DocumentNotFound.new(DataSet, entity: uuid) unless set
      response_from_set set
    end

    post '/:uuid' do
      prevent_access! unless is_authorized_to_access?(uuid)

      create_set(uuid)
    end

    put '/:uuid' do
      prevent_access! unless is_authorized_to_access?(uuid)

      set = DataSet.where(entity: uuid).first
      raise Mongoid::Errors::DocumentNotFound.new(DataSet, entity: uuid) unless set

      set.update_attributes(payload: payload)
      response_from_set set
    end

    put '/:uuid/*key' do
      prevent_access! unless is_authorized_to_access?(uuid)

      set = DataSet.where(entity: uuid).first
      raise Mongoid::Errors::DocumentNotFound.new(DataSet, entity: uuid) unless set

      set.partial_update_attributes(params[:key], payload)
      response_from_set set
    end
  end
end
